tool/src/org/antlr/v4/codegen/target/CSharpTarget.java

```diff
@@ -8,22 +8,119 @@
 import org.antlr.v4.codegen.CodeGenerator;
 import org.antlr.v4.codegen.Target;
 import org.antlr.v4.tool.ErrorType;
-import org.antlr.v4.tool.ast.GrammarAST;
 import org.stringtemplate.v4.NumberRenderer;
 import org.stringtemplate.v4.STErrorListener;
 import org.stringtemplate.v4.STGroup;
 import org.stringtemplate.v4.STGroupFile;
 import org.stringtemplate.v4.StringRenderer;
 import org.stringtemplate.v4.misc.STMessage;
 
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
 public class CSharpTarget extends Target {
+	protected static final HashSet<String> reservedWords = new HashSet<>(Arrays.asList(
+		"abstract",
+		"as",
+		"base",
+		"bool",
+		"break",
+		"byte",
+		"case",
+		"catch",
+		"char",
+		"checked",
+		"class",
+		"const",
+		"continue",
+		"decimal",
+		"default",
+		"delegate",
+		"do",
+		"double",
+		"else",
+		"enum",
+		"event",
+		"explicit",
+		"extern",
+		"false",
+		"finally",
+		"fixed",
+		"float",
+		"for",
+		"foreach",
+		"goto",
+		"if",
+		"implicit",
+		"in",
+		"int",
+		"interface",
+		"internal",
+		"is",
+		"lock",
+		"long",
+		"namespace",
+		"new",
+		"null",
+		"object",
+		"operator",
+		"out",
+		"override",
+		"params",
+		"private",
+		"protected",
+		"public",
+		"readonly",
+		"ref",
+		"return",
+		"sbyte",
+		"sealed",
+		"short",
+		"sizeof",
+		"stackalloc",
+		"static",
+		"string",
+		"struct",
+		"switch",
+		"this",
+		"throw",
+		"true",
+		"try",
+		"typeof",
+		"uint",
+		"ulong",
+		"unchecked",
+		"unsafe",
+		"ushort",
+		"using",
+		"virtual",
+		"values",
+		"void",
+		"volatile",
+		"while"
+	));
+
 	public CSharpTarget(CodeGenerator gen) {
 		super(gen);
 		targetCharValueEscape[0] = "\\0";
 		targetCharValueEscape[0x0007] = "\\a";
 		targetCharValueEscape[0x000B] = "\\v";
 	}
 
+	@Override
+	protected Set<String> getReservedWords() {
+		return reservedWords;
+	}
+
+	@Override
+	protected String escapeWord(String word) {
+		return "@" + word;
+	}
+
+	@Override
+	protected boolean supportsWordEscaping() { return true; }
```

Comment:
why not just have `escapeWord(s)` return `s` instead of having a separate method to test whether we need to escape?

Reply:
Yes, now this method is redundant and should be deleted (but it's required for https://github.com/antlr/antlr4/pull/3453).